### PR TITLE
[Feature] Add ModuleDict inherited from BaseModule and ModuleDict

### PR DESCRIPTION
**Pull request by LutingWang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add ModuleDict base module.

## Modification

mmcv/runner/__init__.py
mmcv/runner/base_module.py
docs/understand_mmcv/cnn.md
docs_zh_CN/understand_mmcv/cnn.md
mmcv/tests/test_runner/test_basemodule.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
Hi @LutingWang , thanks for your contribution. You can merge the master branch to your branch because the circleci of CI depends on a file `.circleci/config.yaml`.
image

In addition, the unit tests should be added and you can refer to `mmcv/tests/test_runner/test_basemodule.py`.
https://github.com/open-mmlab/mmcv/blob/0086bef3e87dbbf5cb4d73d188aa9c92ead22979/tests/test_runner/test_basemodule.py#L523

**Comment by zhouzaida** [comment]
The lint (https://app.circleci.com/pipelines/github/open-mmlab/mmcv/88/workflows/1b121e11-9403-4266-8ffc-ef11b12f257b/jobs/172) failed.
It is recommended to follow the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to install pre-commit which will automatically fix the lint.

File: mmcv/cnn/__init__.py [hunk 1]
```diff
@@ -10,14 +10,15 @@
...
-                    initialize, kaiming_init, normal_init, trunc_normal_init,
-                    uniform_init, xavier_init)
+from .utils import (
+    INITIALIZERS, Caffe2XavierInit, ConstantInit, KaimingInit,
+    NormalInit, PretrainedInit, TruncNormalInit, UniformInit,
+    XavierInit, bias_init_with_prob, caffe2_xavier_init,
+    constant_init, fuse_conv_bn, get_model_complexity_info,
+    initialize, kaiming_init, normal_init, trunc_normal_init,
+    uniform_init, xavier_init)
+# yapf: enable
```
Comment (zhouzaida) [inline mmcv/cnn/__init__.py:21]:
```suggestion
from .utils import (INITIALIZERS, Caffe2XavierInit, ConstantInit, KaimingInit,
                    NormalInit, PretrainedInit, TruncNormalInit, UniformInit,
                    XavierInit, bias_init_with_prob, caffe2_xavier_init,
                    constant_init, fuse_conv_bn, get_model_complexity_info,
                    initialize, kaiming_init, normal_init, trunc_normal_init,
                    uniform_init, xavier_init)
```

Comment (LutingWang) [reply mmcv/cnn/__init__.py]: Sorry, I force-pushed before seeing the recommended modification. It seems that the current version passes the lint check.

Comment (zhouzaida) [reply mmcv/cnn/__init__.py]: Great.

**Review by teamwong111 — APPROVED** [review]
LGTM

File: docs/understand_mmcv/cnn.md [hunk 2]
```diff
@@ -366,17 +366,19 @@ Let us introduce the usage of `initialize` in detail.
...
     `Sequential` is inherited from `BaseModule` and `torch.nn.Sequential`.
 
     `ModuleList` is inherited from `BaseModule` and `torch.nn.ModuleList`.
 
+    `ModuleDict` is inherited from `BaseModule` and `torch.nn.ModuleDict`.
+
     `````python
     import torch.nn as nn
-    from mmcv.runner import BaseModule, Sequential, ModuleList
+    from mmcv.runner import BaseModule, Sequential, ModuleList, ModuleDict
```
Comment (zhouzaida) [inline docs/understand_mmcv/cnn.md:381]: An example of `ModuleDict` should be added here like `ModuleList`.

File: docs_zh_CN/understand_mmcv/cnn.md [hunk 3]
```diff
@@ -355,17 +355,19 @@ conv = ConvModule(
...
     `Sequential` 继承自 `BaseModule` 和 `torch.nn.Sequential`
 
     `ModuleList` 继承自 `BaseModule` 和 `torch.nn.ModuleList`
 
+    `ModuleDict` 继承自 `BaseModule` 和 `torch.nn.ModuleDict`
+
     `````python
     import torch.nn as nn
-    from mmcv.runner import BaseModule, Sequential, ModuleList
+    from mmcv.runner import BaseModule, Sequential, ModuleList, ModuleDict
```
Comment (zhouzaida) [inline docs_zh_CN/understand_mmcv/cnn.md:370]: An example of `ModuleDict` should be added here like `ModuleList`.

File: mmcv/runner/base_module.py [hunk 4]
```diff
@@ -193,3 +193,16 @@ class ModuleList(BaseModule, nn.ModuleList):
     def __init__(self, modules=None, init_cfg=None):
         BaseModule.__init__(self, init_cfg)
         nn.ModuleList.__init__(self, modules)
+
+
+class ModuleDict(BaseModule, nn.ModuleDict):
+    """ModuleDict in openmmlab.
+
+    Args:
+        modules (dict, optional): an dict of modules to add.
```
Comment (MeowZheng) [inline mmcv/runner/base_module.py:202]: a mapping (dictionary) of (string: module) or an iterable of key-value pairs of type (string, module)

**Review by MeowZheng — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**